keps/sig-autoscaling/20191111-autoscaling-with-multiple-thresholds.md

```diff
@@ -0,0 +1,202 @@
+Feel free to check out [the official repository](https://github.com/DataDog/watermarkpodautoscaler) and give us your feedback.
+
+Worth mentioning, we have also included some of the features introduced by @gliush
+and @arjunrn in in the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md) in our controller as we believe those features are must-haves for the next verison of the HPA Controller.
+
+
+## Motivation
+
+- Set a high and a low bound at the resource level to better control autoscaling events.
+
+### Goals
+
+The goal is to improve the ability to customize the parameters triggering a scaling event. We identified 4 vectors of improvement:
+
+- Specifying multiple thresholds on a per HPA basis.
+- Modifying the tolerance (`--horizontal-pod-autoscaler-tolerance`) at the HPA level.
+- Specify forbidden windows to prevent scaling events.
+- Limit the velocity of scaling.
+
+The two last improvements suggestions are tackled in the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md)
+
+### Non-Goals
+
+- This does not aim at replacing the HPA Controller. 
+- Not addressing the features spec'ed in the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md).
+
+## Proposal
+
+Getting the tolerance parameter (`--horizontal-pod-autoscaler-tolerance`) from the controller level to the resource level is a good start, and has been suggested in the past (see [#39090](https://github.com/kubernetes/kubernetes/issues/39090#issuecomment-466398426)). 
+We would like to suggest going beyond this and be able to specify two thresholds. A high bound and a low bound, which extend the idea of a single threshold +/- the tolerance.
+
+There is some overlap between this KEP and the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md), we (Datadog) are heavily relying on the features they introduced and strongly believe they should be made generally available.
+
+### User Stories
+
+We have instrumented the custom controller we use internally to expose metrics, so we can best illustrate the use cases.
+
+#### Canonical use case of the bounds to control the primary scaling condition.
+
+In this example we are using the following Spec configuration:
+```
+    minReplicas: 4
+    maxReplicas: 9
+    metrics:
+    - external:
+        highWatermark: 400m
+        lowWatermark: 150m
+        metricName: custom.request_duration.max
+        metricSelector:
+          matchLabels:
+            kubernetes_cluster: mycluster
+            service: billing
+            short_image: billing-app
+      type: External
+    tolerance: 0.01
```

Comment:
Out of curiousity, what value is tolerance being applied to here?

Reply:
When comparing the recommended replica count to the current replica count.  If the recommended is within 10% of the current, no scaling change is made.

https://github.com/kubernetes/kubernetes/blob/9905a33fd738a52d3f7ae73cf823d82a05720e33/pkg/controller/podautoscaler/replica_calculator.go#L95

Reply:
I was not sure if the KEP was completely accepted, but indeed it seems like it is. Per my [other comment](https://github.com/kubernetes/enhancements/pull/1372#discussion_r350227128) I will update the KEP to be rebased off of the `Configurable Horizontal Pod Autoscaler` one

Reply:
As for the value, I believe that using a "hard" thresholds/bounds logic is not the most ideal given that metrics can be rolled up, aggregated etc depending on the metrics provider, and I still wanted to give the flexibility of having "wiggle room" around those boundaries.

Reply:
@josephburnett probably too late to bikeshed the name, but perhaps "deadband_width" or "deadband_tolerance"?

@CharlyF Is that wiggle-room requirement met by a tolerance measure instead of upper/lower bounds?

Reply:
yes so the new suggestion will encompass using this "buffer" around the target value. In the controller we made, we did add the tolerance around the upper and lower bounds. It was indeed just a tolerance, now it would be more of a `deadband_tolerance`. Although they do represent the same thing down the road.